jax/tree_util.py

```diff
@@ -193,13 +193,29 @@ def build_tree(treedef, xs):
   return treedef.from_iterable_tree(xs)
 
 def tree_transpose(outer_treedef, inner_treedef, pytree_to_transpose):
-  flat, treedef = tree_flatten(pytree_to_transpose)
-  expected_treedef = outer_treedef.compose(inner_treedef)
-  if treedef != expected_treedef:
-    raise TypeError("Mismatch\n{}\n != \n{}".format(treedef, expected_treedef))
+  """Produces a pytree with the same data but a different structure.
+
+  For example, for a ``pytree_to_transpose`` with value
+  ``[dict(a=1, b=2), dict(a=11, b=12)]`` (a list of dictionaries) we can
+  transpose it into a dictionary of lists with value
+  ``dict(a=[1, 11], b=[2, 12])``
+
+  >>> import jax
+  >>> from jax.tree_util import tree_structure, tree_transpose
+  >>>
+  >>> pytree_to_transpose = [dict(a=1, b=2), dict(a=11, b=12)]
+  >>> print(tree_transpose(tree_structure(["*", "*"]),
+  ...                      tree_structure(dict(a="*", b="*")),
+  ...                      pytree_to_transpose))
+  {'a': [1, 11], 'b': [2, 12]}
 
+  """
+  flat, treedef = tree_flatten(pytree_to_transpose)
   inner_size = inner_treedef.num_leaves
   outer_size = outer_treedef.num_leaves
+  if treedef.num_leaves != (inner_size * outer_size):
+    expected_treedef = outer_treedef.compose(inner_treedef)
+    raise TypeError(f"Mismatch\n{treedef}\n != \n{expected_treedef}")
   flat = iter(flat)
   lol = [[next(flat) for _ in range(inner_size)] for __ in range(outer_size)]
```

Comment:
I was wondering why can't this fucntion be implemented as:
```
flat, treedef = tree_flatten(pytree_to_transpose)
new_treedef = inner_treedef.compose(outer_treedef)
return tree_unflatten(new_treedef, flat)
```

Perhaps I do not fully understand its semantics.

Reply:
(Perhaps the original `tree_transpose` implementation predates the `compose` operator on treedefs?)

I thought that might work, so I tried it but got a failure on `APITest.test_jacobian_on_pytrees`. I believe the flat order must change.